locator/topology.cc

```diff
@@ -218,25 +218,25 @@ const node* topology::update_node(node* node, std::optional<host_id> opt_id, std
 
     bool changed = false;
     if (opt_id) {
-        if (*opt_id != node->host_id()) {
+        if (*opt_id != node.host_id()) {
             if (!*opt_id) {
-                on_internal_error(tlogger, seastar::format("Updating node host_id to null is disallowed: {}: new host_id={}", node_printer(node), *opt_id));
+                on_internal_error(tlogger, seastar::format("Updating node host_id to null is disallowed: {}: new host_id={}", node_printer(&node), *opt_id));
             }
-            if (node->is_this_node() && node->host_id()) {
-                on_internal_error(tlogger, seastar::format("This node host_id is already set: {}: new host_id={}", node_printer(node), *opt_id));
+            if (node.is_this_node() && node.host_id()) {
+                on_internal_error(tlogger, seastar::format("This node host_id is already set: {}: new host_id={}", node_printer(&node), *opt_id));
```

Comment:
I see `node_printer` handles the `nullptr` case internally, but I don't know if it's needed now. Maybe we never provide it a pointer that could be `nullptr`. We could check it and perhaps change `node_printer` so that it also stores a reference instead of a pointer and doesn't have to handle the `nullptr` case.

You can try to do it, but I don't think it's necessary. `node_printer` doesn't make reasoning about the code harder even if it uses a pointer.

Reply:
Yes I did modify the node_printer, to work with refs, but noticed some tests failing, I probably could have taken a closer look, but wasn't sure if it was worth it, so left it alone. Let me know if u feel otherwise, thanks